proxy/plugin/cache/base.py

```diff
@@ -27,34 +32,109 @@ class BaseCacheResponsesPlugin(HttpProxyBasePlugin):
     Different storage backends can be used per request if required.
     """
 
+    class EnabledDescriptor:
+        def __init__(self, enabled: bool = False) -> None:
+            self.enabled = multiprocessing.Event()
+            if enabled:
+                self.enabled.set()
+
+        def __get__(self, obj: Optional[object], owner: type) -> Any:
+            if obj is None:
+                return self.enabled
+            return None
+
+    # Dynamically enable / disable cache
+    enabled = EnabledDescriptor(True)
+
     def __init__(
             self,
             *args: Any,
             **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
         self.store: Optional[CacheStore] = None
+        self.__enabled: Optional[bool] = None
 
     def set_store(self, store: CacheStore) -> None:
         self.store = store
 
     def before_upstream_connection(
             self, request: HttpParser) -> Optional[HttpParser]:
+        self.__enabled = self.__class__.enabled.is_set()
+        if not self.__enabled:
+            return request
+
         assert self.store
+        logger.info("Upstream connexion %s:%d %s" %
+                    (text_(request.host), request.port if request.port else 0, text_(request.path)))
+
+        if request.port == 443:
+            return request
+
         try:
-            self.store.open(request)
+            if self.store.is_cached(request):
+                return None
         except Exception as e:
-            logger.info('Caching disabled due to exception message %s', str(e))
+            logger.info('Caching disabled due to exception message: %s', str(e))
+
         return request
 
     def handle_client_request(
             self, request: HttpParser) -> Optional[HttpParser]:
+        assert self.__enabled is not None
+        if not self.__enabled:
+            return request
+
         assert self.store
-        return self.store.cache_request(request)
+        logger.info("Client request %s:%d %s" %
+                    (text_(request.host), request.port if request.port else 0, text_(request.path)))
+
+        if request.port == 443:
```

Comment:
Assuming we have already completed TLS handshake with the client above in `before_upstream_connection` callback, here we can return response from the cache.  There are two possibilities:

1. Save unencrypted response as cache and encrypt it every time while serving from cache.  In-fact, encryption will be automatic when serving clear text response over the TLS connection.
2. Save encrypted response as cache.  Decrypt to parse it as necessary.  Finally serve data over established TLS connection.  This step involves an edge case that can occur when generated server side certificate expires, resulting into inability to decrypt the stored response or in worst case resulting into serving garbage data to client.

Reply:
I would prefer we use solution 1. It seems more practical to me. I sometimes had to look at the cache to understand what data got into it (upsteam has sporadic faults).
That is what I used on https://github.com/pasccom/proxy.py/tree/develop_cache_no_upstream

Reply:
+1 agreed.  Privacy can be argued with solution 1 but this aspect can be controlled by using proper ACL/Permissions over stored cache directories or files.